grpc/go-diatheke/v2/streams.go

```diff
@@ -132,3 +133,134 @@ func WriteTTSAudio(stream *TTSStream, writer io.Writer) error {
 		}
 	}
 }
+
+// TranscribeStream represents a stream of data sent from the client
+// to Diatheke for the purpose of transcribing speech.
+type TranscribeStream struct {
+	PBStream diathekepb.Diatheke_TranscribeClient
+}
+
+// SendAudio to Diatheke for transcription. If the returned error
+// is io.EOF, the stream has already been closed by the server.
+func (ts *TranscribeStream) SendAudio(audio []byte) error {
+	return ts.PBStream.Send(&diathekepb.TranscribeInput{
+		Data: &diathekepb.TranscribeInput_Audio{
+			Audio: audio,
+		},
+	})
+}
+
+// SendAction passes the given transcribe action to Diatheke to use
+// for speech recognition. The action must first be sent before
+// any audio will be transcribed. If the stream was created using
+// client.NewTranscribeStream(), the action was already sent.
+func (ts *TranscribeStream) SendAction(action *diathekepb.TranscribeAction) error {
+	return ts.PBStream.Send(&diathekepb.TranscribeInput{
+		Data: &diathekepb.TranscribeInput_Action{
+			Action: action,
+		},
+	})
+}
+
+// ReceiveResult waits for the next transcribe result from
+// Diatheke. When the returned error is io.EOF, the stream has
+// been closed and no more results will be sent from Diatheke.
+func (ts *TranscribeStream) ReceiveResult() (*diathekepb.TranscribeResult, error) {
+	return ts.PBStream.Recv()
+}
+
+type TranscribeResultHandler func(*diathekepb.TranscribeResult)
+
+// ReadTranscribeAudio is a convenience function to send audio
+// from the given reader to the stream. Results are processed
+// by the given handler as they come. This function will block
+// until the stream ends, either because the server closed it,
+// or the reader returned io.EOF, or there was an error.
+func ReadTranscribeAudio(
+	stream *TranscribeStream,
+	reader io.Reader,
+	buffSize int,
+	handler TranscribeResultHandler,
+) error {
+	// Read audio on a separate go routine
+	sendErrCh := make(chan error)
+	go func() {
+		sendErrCh <- sendTranscribeAudio(stream, reader, buffSize)
+	}()
+
+	// In the meantime, handle results here
+	var receiveErr error
+	for {
+		var result *diathekepb.TranscribeResult
+		result, receiveErr = stream.ReceiveResult()
+		if receiveErr != nil {
+			break
+		}
+
+		handler(result)
+	}
+
+	// Get the error from the send side of the stream
+	sendErr := <-sendErrCh
+
+	if receiveErr == io.EOF {
+		// This just indicates the stream closed. Return a
+		// send error (if there was one).
+		return sendErr
+	}
+
+	// Both of these are real errors, not io.EOF.
+	if sendErr != nil && receiveErr != nil {
+		return fmt.Errorf("send error: %v, receive error: %v",
+			sendErr, receiveErr)
+	}
+
+	if sendErr != nil {
+		return sendErr
+	}
+
+	return receiveErr
+}
+
+// helper function to handle the send side of the transcribe stream.
+func sendTranscribeAudio(
+	stream *TranscribeStream,
+	reader io.Reader,
+	buffSize int,
```

Comment:
```suggestion
	byteBuffSize int,
```

Maybe add a reference to bytes to clarify the size is number of bytes as opposed to number of samples

Reply:
Good idea, I'll update that.